cpp/src/arrow/testing/math.cc

```diff
@@ -0,0 +1,95 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/testing/math.h"
+
+#include <cmath>
+#include <limits>
+
+#include <gtest/gtest.h>
+
+#include "arrow/util/logging.h"
+
+namespace arrow {
+namespace {
+
+template <typename Float>
+constexpr Float kOneUlp;
+template <>
+constexpr float kOneUlp<float> = 5.9604645e-08f;
+template <>
+constexpr double kOneUlp<double> = 1.1102230246251565e-16;
```

Comment:
Are these values `Float(1.0) - std::nextafter(Float(1.0), std::numeric_limits<Float>::min())`?

The names `kOneUlp` could be misleading as it can be also interpreted as `std::nextafter(Float(1.0), std::numeric_limits<Float>::max()) - Float(1.0)` (results in values being twice big).

Reply:
Yes, these are. I realize that the names can be misleading, I can perhaps add a comment.

Reply:
I added a comment, also explaining why we're this definition (because `frexp` returns a mantissa between 0.5 and 1.0 - not between 1.0 and 2.0).